neps/nep-519-yield-execution.md

```diff
@@ -0,0 +1,164 @@
+---
+NEP: 519
+Title: Yield Execution
+Authors: Akhi Singhania <akhi3030@gmail.com>
+Status: Draft
+DiscussionsTo: https://github.com/near/NEPs/pull/519
+Type: Protocol
+Version: 0.0.0
+Created: 2023-11-17
+LastUpdated: 2023-11-20
+---
+
+## Summary
+
+Today, when a smart contract is called by a user or another contract, it has no sensible way to delay responding to the caller.  There exist use cases where contracts would benefit from being able to delay responding till some arbitrary time in the future.  This proposal introduces such possibility into the NEAR protocol.
+
+## Motivation
+
+There exist some situations where when a smart contract on NEAR is called, it will only be able to provide an answer at some arbitrary time in the future.  So the callee needs a way to defer replying to the caller till this time in future.
+
+Examples include when a smart contract (`S`) provides MPC signing capabilities parties external to the NEAR protocol are computing the signature.  The rough steps are:
+1. Signer contract provides a function `fn sign_payload(Payload, ...)`.
+2. When called, the contract updates some contract state which is being monitored by external indexers to indicate that a new signing request has been received.  It also defers replying to the caller.
+3. The indexers observe the new signing request, they compute a signature and call another function `fn signature_available(Signature, ...)` on the signer contract.
+4. The signer contract validates the signature and if validate, replies to the original caller.
+
+Today, the NEAR protocol has no sensible way to defer replying to the caller in step 2 above.  This proposal proposes adding two following new host functions to the NEAR protocol:
+
+- `yield`: this can be called by a contract to indicate to the protocol that it is not ready yet to reply to its caller.
+- `resume`: a contract can use this mechanism to indicate to a protocol that it is now ready to reply to a caller that it had deferred earlier.
+
+If these two host functions were available, then `yield` would be used in step 2 above and `resume` would be used in step 4 above.
+
+## Specification
+
+The proposal is to add the following host functions to the NEAR protocol:
+
+
+```rust
+/// Instructs the protocol that the smart contract is not ready yet to respond
+/// to its caller yet.  The smart contract promises to call `yield_resume()`
+/// within `yield_num_blocks` blocks.  When `yield_resume()` is called, the
+/// protocol will call the method on the smart contract that is identified by
+/// `method_name_len` and `method_name_ptr` and this method will be expected to
+/// either respond to the caller or create another promise.
+///
+/// If the contract fails to call `yield_resume()` within `yield_num_blocks`,
```

Comment:
```suggestion
/// If the contract fails to, in a `yield_resume()`, call `value_return`, `panic` or `abort` within `yield_num_blocks`,
```
This makes the API much more flexible, but does it make it harder to implement?

Reply:
If I understood the idea properly, then I think we need to think about the semantics some more.  Let's say the following has happened:

- contract A called the signer to get a signature and the signer has called yield create.
- contract B called the signer to get a signature and the signer has called yield create.
- now some other contract C calls the signer and the signer `panic`s, which one of the above two outstanding executions should we terminate?  I don't think we can really pick between them so the only sensible choice seems to be to terminate both.  It doesn't seem like that that is a good choice either.

Reply:
So if contract C calls the signer and that call panics before it calls `yield_create` then the call from contract C should panic and none of the others.

This refers to a situation where contract D calls `yield_create`, the signer calls `yield_resume` and then the resumed call panics, ending contract D's suspended call. It shouldn't effect any other outstanding calls.